File: retrocpu_boot_monitor/mn1613/test/mn1613_mon_settings.ts
/**
 * mn1613_mon のリンク設定（Makefile の ASM_SRCS と同じ）。
 * 根拠: asm_test_framework.mdc / retrocpu_boot_monitor/Makefile / emulater_code_test.mdc
 */
import path from "node:path";
import { fileURLToPath } from "node:url";
import type {
  CodeTestIoMockEntry,
  CpuLogMode,
  JsonTestSettings,
} from "../../../retrocpu_test_framework/src/index.js";

/** `mn1613/logs`（テスト専用 CPU ログ） */
export const MN1613_TEST_LOG_DIR = path.resolve(
  path.dirname(fileURLToPath(import.meta.url)),
  "../logs",
);

/**
 * `*_test.ts` に対応する asm 名の `.log` を `cpuLogFile` に付ける。
 * `handshake_timer_test.ts` → `mn1613/logs/handshake_timer.log`
 * @param settings 元設定
 * @param testMetaUrl 呼び出し元の `import.meta.url`
 * @param opts 任意。`cpuLogMode` など
 * @returns cpuLogFile 付き設定
 */
export function withMn1613CpuLog(
  settings: JsonTestSettings,
  testMetaUrl: string,
  opts?: { cpuLogMode?: CpuLogMode },
): JsonTestSettings {
  const envMode = process.env.MN1613_CPU_LOG_MODE;
  const envCpuLogMode: CpuLogMode | undefined =
    envMode === "checkpoint" || envMode === "instruction" ? envMode : undefined;
  const selectedMode = opts?.cpuLogMode ?? envCpuLogMode;
  const stem = path
    .basename(fileURLToPath(testMetaUrl))
    .replace(/_test\.ts$/i, "");
  return {
    ...settings,
    cpuLogFile: path.join(MN1613_TEST_LOG_DIR, `${stem}.log`),
    ...(selectedMode ? { cpuLogMode: selectedMode } : {}),
  };
}

/**
 * モニタ結合テスト共通。
 * 入力は Makefile 成果物の `.ihx` / `.cdb` のみ（`.asm` は読まない）。
 */
export const mn1613MonSettings: JsonTestSettings = {
  name: "mn1613_mon",
  cpu: "mn1613",
  hexFile: "${MONITOR_HEX}/mn1613_mon.ihx",
  cdbFile: "${MONITOR_HEX}/mn1613_mon.cdb",
  initLabel: "g_main",
};

/** ハンドシェイク結合テスト用 ioMock（emulater_code_test.mdc §7） */
export const mn1613MonHandshakeIoMock: CodeTestIoMockEntry[] = [
  { type: "handshake", timeoutMs: 5000, syncIrq2: false },
];

/** HEX/CDB + handshake ioMock。reload で RD/WT をキックする */
export const mn1613MonHandshakeSettings: JsonTestSettings = {
  ...mn1613MonSettings,
  ioMock: mn1613MonHandshakeIoMock,
};

File: retrocpu_boot_monitor/test/mn1613/mn1613_mon_settings.rs
use std::path::{Path, PathBuf};

use retrocpu_test_framework_rs::{
    create_session_from_settings, AsmCpuType, CodeTestIoMockEntry, CpuLogMode, FrameworkError,
    JsonTestSettings,
};

fn derive_log_stem(test_file: &str) -> String {
    let p = Path::new(test_file);
    let file = p.file_name().and_then(|s| s.to_str()).unwrap_or(test_file);
    file.trim_end_matches("_test.ts")
        .trim_end_matches("_test.rs")
        .to_string()
}

fn select_cpu_log_mode(explicit: Option<CpuLogMode>, env_mode: Option<&str>) -> Option<CpuLogMode> {
    if explicit.is_some() {
        return explicit;
    }
    match env_mode {
        Some("checkpoint") => Some(CpuLogMode::Checkpoint),
        Some("instruction") => Some(CpuLogMode::Instruction),
        _ => None,
    }
}

fn with_mn1613_cpu_log_rs(
    settings: &JsonTestSettings,
    test_file: &str,
    explicit_mode: Option<CpuLogMode>,
    env_mode: Option<&str>,
) -> JsonTestSettings {
    let mut s = settings.clone();
    let repo = super::repo_root();
    let stem = derive_log_stem(test_file);
    let log_path = repo
        .join("retrocpu_boot_monitor")
        .join("logs")
        .join("mn1613")
        .join(format!("{stem}.log"));
    s.cpu_log_file = Some(log_path.to_string_lossy().to_string());
    s.cpu_log_mode = select_cpu_log_mode(explicit_mode, env_mode);
    s
}

fn mn1613_mon_handshake_rs_settings() -> JsonTestSettings {
    let mut s = super::mn1613_rs_settings();
    s.io_mock = Some(vec![CodeTestIoMockEntry::Handshake]);
    s
}

#[test]
fn mn1613_mon_settings_points_to_rs_artifacts() {
    let s = super::mn1613_rs_settings();
    let (hex, cdb, log) = super::mn1613_monitor_paths();
    assert_eq!(s.name, "mn1613_mon_rs");
    assert_eq!(s.cpu, AsmCpuType::Mn1613);
    assert_eq!(PathBuf::from(&s.hex_file), hex);
    assert_eq!(PathBuf::from(&s.cdb_file), cdb);
    assert_eq!(PathBuf::from(s.cpu_log_file.expect("cpu_log_file")), log);
    assert_eq!(s.init_label.as_deref(), Some("g_main"));
}

#[test]
fn with_mn1613_cpu_log_rs_sets_log_file_stem_from_test_name() {
    let base = super::mn1613_rs_settings();
    let s = with_mn1613_cpu_log_rs(
        &base,
        "test/mn1613/handshake/handshake_timer_test.rs",
        None,
        None,
    );

    let p = PathBuf::from(s.cpu_log_file.expect("cpu_log_file"));
    assert_eq!(
        p.file_name().and_then(|x| x.to_str()),
        Some("handshake_timer.log")
    );
    assert!(p
        .to_string_lossy()
        .contains("retrocpu_boot_monitor/logs/mn1613/handshake_timer.log"));
}

#[test]
fn with_mn1613_cpu_log_rs_prefers_explicit_mode_over_env() {
    let base = super::mn1613_rs_settings();
    let s = with_mn1613_cpu_log_rs(
        &base,
        "test/mn1613/handshake/handshake_timer_test.rs",
        Some(CpuLogMode::Instruction),
        Some("checkpoint"),
    );
    assert_eq!(s.cpu_log_mode, Some(CpuLogMode::Instruction));
}

#[test]
fn with_mn1613_cpu_log_rs_uses_valid_env_mode_when_explicit_absent() {
    let base = super::mn1613_rs_settings();
    let checkpoint = with_mn1613_cpu_log_rs(
        &base,
        "test/mn1613/handshake/handshake_timer_test.rs",
        None,
        Some("checkpoint"),
    );
    let instruction = with_mn1613_cpu_log_rs(
        &base,
        "test/mn1613/handshake/handshake_timer_test.rs",
        None,
        Some("instruction"),
    );
    let invalid = with_mn1613_cpu_log_rs(
        &base,
        "test/mn1613/handshake/handshake_timer_test.rs",
        None,
        Some("other"),
    );

    assert_eq!(checkpoint.cpu_log_mode, Some(CpuLogMode::Checkpoint));
    assert_eq!(instruction.cpu_log_mode, Some(CpuLogMode::Instruction));
    assert_eq!(invalid.cpu_log_mode, None);
}

#[test]
fn handshake_settings_attach_handshake_mock_and_session_uses_it() -> Result<(), FrameworkError> {
    let mut s = create_session_from_settings(&mn1613_mon_handshake_rs_settings(), None)?;
    let hs = s.require_handshake_mock()?;
    hs.set_timestamp_u64(0x0123_4567_89ab_cdef);
    s.detach_io_mock();
    Ok(())
}
